tests/expr/test_expr.py

```diff
@@ -88,7 +88,14 @@ def test_abs():
 
 @pytest.mark.parametrize(("veganame", "methodname"), _remap_classmethod_names(expr))
 def test_expr_funcs(veganame: str, methodname: str):
-    """Test all functions defined in expr.funcs."""
+    """
+    Test all functions defined in expr.funcs.
+
+    # FIXME: These tests are no longer suitable
+    They only work for functions with a **single** argument:
+
+        TypeError: expr.if_() missing 2 required positional arguments: 'thenValue' and 'elseValue'.
+    """
```

Comment:
# FIXME
The tests produced by `test_expr_funcs` are no longer valid.

They only work for functions with a **single** argument:

```py
>>> alt.expr.if_(alt.datum.xxx)
TypeError: expr.if_() missing 2 required positional arguments: 'thenValue' and 'elseValue'.
```

This is **to be expected**, and was demonstrated in the examples of https://github.com/vega/altair/issues/3563

---

Solving this will likely use `inspect.signature` to produce new tests that account for the number of arguments required.

Reply:
Partially resolved in: [test: Adds test_dummy_expr_funcs](https://github.com/vega/altair/pull/3600/commits/9238fb635c37c3527851a3f59244ff71fe1571b2)

Still need to replace `expr.__init__`

Reply:
For comparison, these were all the tests that would fail:
<details>
<summary><b> 62 failed tests</b></summary>

```py
tests\expr\test_expr.py:138: TypeError
============== short test summary info ==============
FAILED tests/expr/test_expr.py::test_expr_funcs[bandspace-bandspace] - AssertionError: assert 'bandspace(da...xx,null,null)' == 'bandspace(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[bandwidth-bandwidth] - AssertionError: assert 'bandwidth(datum.xxx,null)' == 'bandwidth(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[range-range] - AssertionError: assert 'range(datum.xxx,null)' == 'range(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[clampRange-clampRange] - TypeError: expr.clampRange() missing 2 required positional arguments: 'min' and 'max'
FAILED tests/expr/test_expr.py::test_expr_funcs[regexp-regexp] - AssertionError: assert 'regexp(datum.xxx,null)' == 'regexp(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[copy-copy] - AssertionError: assert 'copy(datum.xxx,null)' == 'copy(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[debug-debug] - AssertionError: assert 'debug(datum.xxx,null)' == 'debug(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[replace-replace] - TypeError: expr.replace() missing 2 required positional arguments: 'pattern' and 'replacement'
FAILED tests/expr/test_expr.py::test_expr_funcs[sampleLogNormal-sampleLogNormal] - AssertionError: assert 'sampleLogNor...tum.xxx,null)' == 'sampleLogNormal(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[domain-domain] - AssertionError: assert 'domain(datum.xxx,null)' == 'domain(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[sampleNormal-sampleNormal] - AssertionError: assert 'sampleNormal(datum.xxx,null)' == 'sampleNormal(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[scale-scale] - TypeError: expr.scale() missing 1 required positional argument: 'value'
FAILED tests/expr/test_expr.py::test_expr_funcs[sampleUniform-sampleUniform] - AssertionError: assert 'sampleUnifor...tum.xxx,null)' == 'sampleUniform(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[slice-slice] - TypeError: expr.slice() missing 1 required positional argument: 'start'
FAILED tests/expr/test_expr.py::test_expr_funcs[test-test] - AssertionError: assert 'test(datum.xxx,null)' == 'test(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[substring-substring] - TypeError: expr.substring() missing 1 required positional argument: 'start'
FAILED tests/expr/test_expr.py::test_expr_funcs[timeOffset-timeOffset] - TypeError: expr.timeOffset() missing 1 required positional argument: 'date'
FAILED tests/expr/test_expr.py::test_expr_funcs[atan2-atan2] - TypeError: expr.atan2() missing 1 required positional argument: 'dx'
FAILED tests/expr/test_expr.py::test_expr_funcs[timeSequence-timeSequence] - TypeError: expr.timeSequence() missing 2 required positional arguments: 'start' and 'stop'
FAILED tests/expr/test_expr.py::test_expr_funcs[timeFormat-timeFormat] - TypeError: expr.timeFormat() missing 1 required positional argument: 'specifier'
FAILED tests/expr/test_expr.py::test_expr_funcs[timeParse-timeParse] - TypeError: expr.timeParse() missing 1 required positional argument: 'specifier'
FAILED tests/expr/test_expr.py::test_expr_funcs[clamp-clamp] - TypeError: expr.clamp() missing 2 required positional arguments: 'min' and 'max'
FAILED tests/expr/test_expr.py::test_expr_funcs[datetime-datetime] - TypeError: expr.datetime() missing 1 required positional argument: 'month'
FAILED tests/expr/test_expr.py::test_expr_funcs[timeUnitSpecifier-timeUnitSpecifier] - AssertionError: assert 'timeUnitSpec...tum.xxx,null)' == 'timeUnitSpecifier(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[treeAncestors-treeAncestors] - TypeError: expr.treeAncestors() missing 1 required positional argument: 'node'
FAILED tests/expr/test_expr.py::test_expr_funcs[truncate-truncate] - TypeError: expr.truncate() missing 1 required positional argument: 'length'
FAILED tests/expr/test_expr.py::test_expr_funcs[info-info] - AssertionError: assert 'info(datum.xxx,null)' == 'info(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[utcFormat-utcFormat] - TypeError: expr.utcFormat() missing 1 required positional argument: 'specifier'
FAILED tests/expr/test_expr.py::test_expr_funcs[inrange-inrange] - TypeError: expr.inrange() missing 1 required positional argument: 'range'
FAILED tests/expr/test_expr.py::test_expr_funcs[treePath-treePath] - TypeError: expr.treePath() missing 2 required positional arguments: 'source' and 'target'
FAILED tests/expr/test_expr.py::test_expr_funcs[utc-utc] - TypeError: expr.utc() missing 1 required positional argument: 'month'
FAILED tests/expr/test_expr.py::test_expr_funcs[utcOffset-utcOffset] - TypeError: expr.utcOffset() missing 1 required positional argument: 'date'
FAILED tests/expr/test_expr.py::test_expr_funcs[utcParse-utcParse] - TypeError: expr.utcParse() missing 1 required positional argument: 'specifier'
FAILED tests/expr/test_expr.py::test_expr_funcs[utcSequence-utcSequence] - TypeError: expr.utcSequence() missing 2 required positional arguments: 'start' and 'stop'
FAILED tests/expr/test_expr.py::test_expr_funcs[warn-warn] - AssertionError: assert 'warn(datum.xxx,null)' == 'warn(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[zoomLinear-zoomLinear] - TypeError: expr.zoomLinear() missing 2 required positional arguments: 'anchor' and 'scaleFactor'
FAILED tests/expr/test_expr.py::test_expr_funcs[zoomLog-zoomLog] - TypeError: expr.zoomLog() missing 2 required positional arguments: 'anchor' and 'scaleFactor'
FAILED tests/expr/test_expr.py::test_expr_funcs[gradient-gradient] - TypeError: expr.gradient() missing 2 required positional arguments: 'p0' and 'p1'
FAILED tests/expr/test_expr.py::test_expr_funcs[zoomPow-zoomPow] - TypeError: expr.zoomPow() missing 3 required positional arguments: 'anchor', 'scaleFactor', and 'exponent'
FAILED tests/expr/test_expr.py::test_expr_funcs[zoomSymlog-zoomSymlog] - TypeError: expr.zoomSymlog() missing 3 required positional arguments: 'anchor', 'scaleFactor', and 'constant'
FAILED tests/expr/test_expr.py::test_expr_funcs[geoArea-geoArea] - TypeError: expr.geoArea() missing 1 required positional argument: 'feature'
FAILED tests/expr/test_expr.py::test_expr_funcs[geoBounds-geoBounds] - TypeError: expr.geoBounds() missing 1 required positional argument: 'feature'
FAILED tests/expr/test_expr.py::test_expr_funcs[format-format] - TypeError: expr.format() missing 1 required positional argument: 'specifier'
FAILED tests/expr/test_expr.py::test_expr_funcs[indata-indata] - TypeError: expr.indata() missing 2 required positional arguments: 'field' and 'value'
FAILED tests/expr/test_expr.py::test_expr_funcs[invert-invert] - TypeError: expr.invert() missing 1 required positional argument: 'value'
FAILED tests/expr/test_expr.py::test_expr_funcs[geoCentroid-geoCentroid] - TypeError: expr.geoCentroid() missing 1 required positional argument: 'feature'
FAILED tests/expr/test_expr.py::test_expr_funcs[indexof-indexof] - TypeError: expr.indexof() missing 1 required positional argument: 'value'
FAILED tests/expr/test_expr.py::test_expr_funcs[lastindexof-lastindexof] - TypeError: expr.lastindexof() missing 1 required positional argument: 'value'
FAILED tests/expr/test_expr.py::test_expr_funcs[if-if_] - TypeError: expr.if_() missing 2 required positional arguments: 'thenValue' and 'elseValue'
FAILED tests/expr/test_expr.py::test_expr_funcs[lerp-lerp] - TypeError: expr.lerp() missing 1 required positional argument: 'fraction'
FAILED tests/expr/test_expr.py::test_expr_funcs[panSymlog-panSymlog] - TypeError: expr.panSymlog() missing 2 required positional arguments: 'delta' and 'constant'
FAILED tests/expr/test_expr.py::test_expr_funcs[min-min] - TypeError: expr.min() missing 1 required positional argument: 'value2'
FAILED tests/expr/test_expr.py::test_expr_funcs[pluck-pluck] - TypeError: expr.pluck() missing 1 required positional argument: 'field'
FAILED tests/expr/test_expr.py::test_expr_funcs[geoScale-geoScale] - AssertionError: assert 'geoScale(datum.xxx,null)' == 'geoScale(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[pad-pad] - TypeError: expr.pad() missing 1 required positional argument: 'length'
FAILED tests/expr/test_expr.py::test_expr_funcs[join-join] - AssertionError: assert 'join(datum.xxx,null)' == 'join(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[merge-merge] - AssertionError: assert 'merge(datum.xxx,null)' == 'merge(datum.xxx)'
FAILED tests/expr/test_expr.py::test_expr_funcs[max-max] - TypeError: expr.max() missing 1 required positional argument: 'value2'
FAILED tests/expr/test_expr.py::test_expr_funcs[panPow-panPow] - TypeError: expr.panPow() missing 2 required positional arguments: 'delta' and 'exponent'
FAILED tests/expr/test_expr.py::test_expr_funcs[pow-pow] - TypeError: expr.pow() missing 1 required positional argument: 'exponent'
FAILED tests/expr/test_expr.py::test_expr_funcs[panLog-panLog] - TypeError: expr.panLog() missing 1 required positional argument: 'delta'
FAILED tests/expr/test_expr.py::test_expr_funcs[panLinear-panLinear] - TypeError: expr.panLinear() missing 1 required positional argument: 'delta'
============== 62 failed, 251 passed in 14.38s ==============
```

</details>

Reply:
Resolved in https://github.com/vega/altair/pull/3600#commits-pushed-7311fdf